src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
+
+    def report_error(self, err: int) -> None:
+        if err == errno.ENOENT:
+            problem = 'not-found'
+        elif err in (errno.EPERM, errno.EACCES):
+            problem = 'access-denied'
+        elif err == errno.ENOTDIR:
+            problem = 'not-directory'
+        else:
+            problem = 'internal-error'
+
+        self.send_update({'error': {
+            'problem': problem, 'message': os.strerror(err), 'errno': errno.errorcode[err]
+        }}, reset=True)
+
+    def flag_onlydir_error(self, fd: Handle) -> bool:
+        # If our requested path ended with '/' then make sure we got a
+        # directory, or else it's an error.  open() will have already flagged
+        # that for us, but systemd_ctypes doesn't do that (yet).
+        if not self.watch or not self.path.endswith('/'):
+            return False
+
+        buf = os.fstat(fd)  # this should never fail
+        if stat.S_IFMT(buf.st_mode) != stat.S_IFDIR:
+            self.report_error(errno.ENOTDIR)
+            return True
+
+        return False
+
+    def report_initial_state(self, fd: Handle) -> None:
+        if self.flag_onlydir_error(fd):
+            return
+
+        self.fd = fd
+
+        entries = {''}
+        if self.fnmatch:
+            try:
+                entries.update(os.listdir(f'/proc/self/fd/{self.fd}'))
+                self.effective_fnmatch = self.fnmatch
+            except OSError:
+                # If we failed to get an initial list, then report nothing from now on
+                self.effective_fnmatch = ''
+
+        self.process_update({e for e in entries if self.interesting(e)}, reset=True)
+
+    def do_inotify_event(self, mask: InotifyEvent, cookie: int, rawname: 'bytes | None') -> None:
+        logger.debug('do_notify_event(%r, %r, %r)', mask, cookie, rawname)
+        name = (rawname or b'').decode(errors='surrogateescape')
+
+        self.schedule_update(name)
+
+        if name and mask | (InotifyEvent.CREATE | InotifyEvent.DELETE |
+                            InotifyEvent.MOVED_TO | InotifyEvent.MOVED_FROM):
+            # These events change the mtime of the directory
+            self.schedule_update('')
+
+    def do_identity_changed(self, fd: 'Handle | None', err: 'int | None') -> None:
+        # If there were previously pending changes, they are now irrelevant.
+        if self.pending is not None:
+            # Note: don't set to None, since the handler is still pending
+            self.pending.clear()
+
+        if err is None:
+            assert fd is not None
+            self.report_initial_state(fd)
+        else:
+            self.report_error(err)
+
+    def do_close(self) -> None:
+        if self.path_watch is not None:
+            self.path_watch.close()
+        self.close()
+
+    def do_open(self, options: JsonObject) -> None:
+        self.path = get_str(options, 'path')
+        if not os.path.isabs(self.path):
+            raise JsonError(options, '"path" must be an absolute path')
```

Comment:
That's sensible as it avoids the assumption that the bridge runs in ~. However, should this support `os.path.expanduser()`, so that callers can watch e.g. `~/.ssh`? That seems useful to me. Totally fine as follow-up of course if/when we need it, just gauging your opinion.

Reply:
I don't think we support this on any of the other filesystem channels...  My initial instinct here would to be complain about "what if the filename happens to start with `~`?", but indeed, it needs to start with `/`, so we're not in danger here...

If we wanted to add that as a feature, we should add it everywhere, and in a separate commit, I'd say.